linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java

```diff
@@ -90,12 +111,10 @@ static Bom copyWithSnapshot(RepositorySystem repositorySystem, Bom bom)
     RepositorySystemSession session = RepositoryUtility.newSession(repositorySystem);
 
     for (Artifact managedDependency : bom.getManagedDependencies()) {
-      Optional<String> snapshotVersion =
-          findSnapshotVersion(repositorySystem, session, managedDependency);
-      if (snapshotVersion.isPresent()) {
-        managedDependency = managedDependency.setVersion(snapshotVersion.get());
-      }
-      managedDependencies.add(managedDependency);
+      managedDependencies.add(
+          findSnapshotVersion(repositorySystem, session, managedDependency)
+              .map(managedDependency::setVersion)
+              .orElse(managedDependency));
```

Comment:
This is from David's comment in #694

Reply:
Ha! Now I think you're going a bit further than you need to.

`managedDependency.setVersion(foo)` returns its receiver, so this works, but both the `map` and the `orelse` are actually returning the same object, and that is not what I expect when seeing `orElse`.

I recommend

```java
findSnapshotVersion(repositorySystem, session, managedDependency)
    .ifPresent(managedDependency::setVersion);
managedDependencies.add(managedDependency);
````

Alternatively, I guess you could change `findSnapshotVersion` into `applySnapshotVersion`, and have it set the version if it finds it. Then you'd have:

```java
applySnapshotVersion(repositorySystem, session, managedDependency);
managedDependencies.add(managedDependency);
````
Up to you between those.

Reply:
`Artifact` is designed as immutable.

```
/**
 * A specific artifact. In a nutshell, an artifact has identifying coordinates and optionally a file that denotes its
 * data. <em>Note:</em> Artifact instances are supposed to be immutable, e.g. any exposed mutator method returns a new
 * artifact instance and leaves the original instance unchanged. <em>Note:</em> Implementors are strongly advised to
 * inherit from {@link AbstractArtifact} instead of directly implementing this interface.
 * 
 * @noimplement This interface is not intended to be implemented by clients.
 * @noextend This interface is not intended to be extended by clients.
 */
public interface Artifact

```

AbstractArtifact:

```
    /**
     * Creates a new artifact with the specified coordinates, properties and file.
     * 
     * @param version The version of the artifact, may be {@code null}.
     * @param properties The properties of the artifact, may be {@code null} if none. The method may assume immutability
     *            of the supplied map, i.e. need not copy it.
     * @param file The resolved file of the artifact, may be {@code null}.
     * @return The new artifact instance, never {@code null}.
     */
    private Artifact newInstance( String version, Map<String, String> properties, File file )
    {
        return new DefaultArtifact( getGroupId(), getArtifactId(), getClassifier(), getExtension(), version, file,
                                    properties );
    }

    public Artifact setVersion( String version )
    {
        String current = getVersion();
        if ( current.equals( version ) || ( version == null && current.length() <= 0 ) )
        {
            return this;
        }
        return newInstance( version, getProperties(), getFile() );
    }
```

Reply:
Oof. "setVersion" is a bad name for that method. OK.